### PR TITLE
Unroll and optimize

### PR DESCRIPTION
**Pull request by j2kun**
Main part of https://github.com/google/heir/issues/257

### CODE REVIEW

**Comment by j2kun** [comment]
Still TODO:

- Get an end-to-end example working with a loop nest that is unrolled and optimized
- Print stats to the terminal
- Figure out a heuristic for stopping

**Comment by j2kun** [comment]
The merging of adjacent generics fails because the loop unrolling inserts `affine.apply` instructions between the generics

```mlir
affine.for %arg2 = 0 to 10 step 2 {
  secret.generic ins(%arg0, %0 : !secret.secret<memref<10x10xi8>>, !secret.secret<memref<1xi8>>) {
  ^bb0(%arg3: memref<10x10xi8>, %arg4: memref<1xi8>):
    %2 = affine.load %arg3[%arg1, %arg2] : memref<10x10xi8>
    %3 = affine.load %arg4[0] : memref<1xi8>
    %4 = arith.addi %2, %3 : i8
    affine.store %4, %arg4[0] : memref<1xi8>
    secret.yield
  }
  %1 = affine.apply affine_map<(d0) -> (d0 + 1)>(%arg2)
  secret.generic ins(%arg0, %0 : !secret.secret<memref<10x10xi8>>, !secret.secret<memref<1xi8>>) {
  ^bb0(%arg3: memref<10x10xi8>, %arg4: memref<1xi8>):
    %2 = affine.load %arg3[%arg1, %1] : memref<10x10xi8>
    %3 = affine.load %arg4[0] : memref<1xi8>
    %4 = arith.addi %2, %3 : i8
    affine.store %4, %arg4[0] : memref<1xi8>
    secret.yield
  }
}
```

**Comment by j2kun** [comment]
After rebasing, now it's complaining that `Only functions with a single return type are supported`, and so it can't emit metadata. Tomorrow I will have to think about how to bypass this, since in our case the arguments are memrefs that are written to as outparams.

**Comment by j2kun** [comment]
Making incremental progress here:

- Supporting zero return types in VerilogEmitter and EmitMetadata, because some of the generic bodies of the affine loops don't return anything (they just write to memrefs which are inputs).


Next step seems to be updating the post-yosys step `convertOpOperands` that adds a compatibility layer between the original input type and the yosys-converted input type. I think I may just expand the scope of `secret.cast` to cover memref conversions.

**Comment by j2kun** [comment]
I ended up removing an assert that now results in the "compatibility layer" simply being `secret.cast` a `secret<memref<5x4xi8>>` to a `secret<tensor<160xi1>>` and back again, since I'm expecting @asraa to switch from tensor to memref in this code path anyway.

I also decided to expand affine.apply (which is inserted during loop unrolling) and replace it with standard arith ops, but now it runs into the problem that the inputs to the `secret.generic` are `index` type, and that is not supported in the VerilogEmitter. I was trying to remember how we handled this before... just pick a bit width and insert cast ops?

**Comment by j2kun** [comment]
After a bit more progress, now I'm blocked on the fact that the design output by Yosys/ABC has this cell whose output write is the same as one of its input wires

```
  attribute \LUT 4'0110
  attribute \module_not_derived 1
  attribute \src "lib/Transforms/YosysOptimizer/yosys/map_lut_to_lut3.v:47.10-56.19"
  cell \lut3 \_1686_
    connect \A $abc$85582$new_n1096_
    connect \B \arg2 [0]
    connect \C 0
    connect \P0 1'0
    connect \P1 1'1
    connect \P2 1'1
    connect \P3 1'0
    connect \P4 0
    connect \P5 0
    connect \P6 0
    connect \P7 0
    connect \Y \arg2 [0]
  end
```

??? How is this possible?

**Comment by j2kun** [comment]
OK, I've taken the approach of lifting the loads and stores out of the generic to optimize. Still TODO:

- Support for > 1 return value in yosys optimizer
- Finish the test I'm currently using to drive the feature work
- Add new tests:
  - Make a loop for which, when you unroll it you can actually optimize better than if you don't
  - Make a loop which has per-iteration dependent values, and add a forward store to load pattern

**Comment by j2kun** [comment]
It works!!! My god, I'm free.

@asraa this is currently branched off https://github.com/google/heir/pull/413, and after that goes in I'll rebase this and clean up the commits a bit.

**Comment by j2kun** [comment]
Added a more complicated test that computes a cumulative sum of an array, however, the central generic op seems to be sometimes split into multiple, bypassing the joint optimization. So I will have to pick this up tomorrow on that thread.

**Comment by j2kun** [comment]
This PR still needs to be rebased over https://github.com/google/heir/pull/413

**Comment by j2kun** [comment]
Rebased and squashed: @asraa it should be ready for review!

File: include/Transforms/YosysOptimizer/YosysOptimizer.td [hunk 1]
```diff
@@ -23,4 +35,22 @@ def YosysOptimizer : Pass<"yosys-optimizer"> {
   ];
 }
 
+def UnrollAndOptimizeAnalysis : Pass<"unroll-and-optimize-analysis"> {
```
Comment (j2kun) [inline include/Transforms/YosysOptimizer/YosysOptimizer.td:38]: To be fair, this part is not yet done. I will work on this in a followup PR.

File: lib/Transforms/YosysOptimizer/YosysOptimizer.cpp [hunk 2]
```diff
@@ -121,8 +132,10 @@ Value convertIntegerValue(Value value, Type convertedType, OpBuilder &b,
   return allocOp.getResult();
 }
 
-/// Convert a secret.generic's operands secret.secret<i3>
-/// to secret.secret<memref<3xi1>>.
+/// Convert a secret.generic's operands:
+///
+///  - secret.secret<i3> -> secret.secret<tensor<3xi1>>
+///  - secret.secret<memref<Nxi3>> -> secret.secret<memref<3Nxi1>>
```
Comment (asraa) [inline lib/Transforms/YosysOptimizer/YosysOptimizer.cpp:138]: are these accurate? i think we generally are converting `i3` to `memref<3xi1>` and `memref<Nxi3>` to `memref<Nx3xi1>`

**Review by asraa — APPROVED** [review]

File: lib/Transforms/YosysOptimizer/YosysOptimizer.cpp (hunk 2 above)
Comment (j2kun) [reply lib/Transforms/YosysOptimizer/YosysOptimizer.cpp:138]: Will revert. I'm not sure why I wrote this.

File: include/Transforms/YosysOptimizer/YosysOptimizer.td (hunk 1 above)
Comment (j2kun) [reply include/Transforms/YosysOptimizer/YosysOptimizer.td:38]: Added TODO

**Comment by j2kun** [comment]
Merged via f6472d488a8fc9b81e9e2a6a8e7848eebabd3088

**Pull request closed without merging**